# Step 1: Load the Dataset

Do the following under Step 1,
1. Download the csv file
2. Upload the csv file in Colab
3. Load the dataset

In [1]:
# Import the pandas module
import pandas as pd

# Set the uploaded dataset in a variable
# Use read_csv function to read the uploaded csv file and display it in a tabular format
data = pd.read_csv("/content/mobile_price_classification.csv")

data.head() # This will show first few records of the dataset

,battery_power,bluetooth,clock_speed,dual_sim,front_cam,4G,int_memory,m_dep,mobile_wt,n_cores,...,px_height,px_width,ram,sc_h,sc_w,talk_time,three_g,touch_screen,wifi,price_range
0,842,0,2.2,0,1,0,7,0.6,188,2,...,20,756,2549,9,7,19,0,0,1,1
1,1021,1,0.5,1,0,1,53,0.7,136,3,...,905,1988,2631,17,3,7,1,1,0,2
2,563,1,0.5,1,2,1,41,0.9,145,5,...,1263,1716,2603,11,2,9,1,1,0,2
3,615,1,2.5,0,0,0,10,0.8,131,6,...,1216,1786,2769,16,8,11,1,0,0,2
4,1821,1,1.2,0,13,1,44,0.6,141,2,...,1208,1212,1411,8,2,15,1,1,0,1


# Step 2: Import Required Libraries

In [2]:
# Import required libraries

import pandas as pd                                                 # For data handling
import numpy as np                                                  # For numerical operations

from sklearn.model_selection import train_test_split                # For splitting data
from sklearn.preprocessing import StandardScaler                    # For feature scaling
from sklearn.metrics import accuracy_score, classification_report   # For metrics

import tensorflow as tf                                             # TensorFlow library for deep learning
from tensorflow.keras.models import Sequential                      # For ANN model
from tensorflow.keras.layers import Dense                           # For ANN layers
from tensorflow.keras.optimizers import Adam                        # Optimizer

In [3]:
# Verify TensorFlow version
print(tf.__version__)

2.19.0


In [4]:
# This I tried, as I was getting warnings while importing the modules from tensorflow
!pip install tensorflow

# Step 3: Separate Features and Label

In [5]:
# Separate input features (X) and output label (y)

X = data.drop("price_range", axis=1)   # All columns except price_range
y = data["price_range"]                # Target column

# Step 4: Train-Test Split

In [6]:
# Split data into training and testing sets

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# For training 80% data
# For Testing 20% data

# Step 5: Feature Scaling

In [7]:
# Scale the features

scaler = StandardScaler()

X_train = scaler.fit_transform(X_train)   # Fit on training data
X_test = scaler.transform(X_test)         # Transform test data

# Here one-hot encoding target labels using LabelBinarizer is not required because,
1. price_range is already a classification label (0,1,2,3)
2. These are already encoded class labels
3. It is Multi-Class Classification
4. The model understands them directly

# Step 6: Build the ANN Model

In [8]:
# Create ANN model

model = Sequential()

# Input Layer &  Hidden Layer 1
model.add(Dense(units=64, activation='relu', input_dim=X_train.shape[1]))

# Hidden Layer 2
model.add(Dense(units=32, activation='relu'))

# Output Layer (with 4 classes)
model.add(Dense(units=4, activation='softmax'))

/usr/local/lib/python3.12/dist-packages/keras/src/layers/core/dense.py:93: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


# Step 7: Compile the Model

In [9]:
# Compile the model

model.compile(
    optimizer=Adam(learning_rate=0.001), loss='sparse_categorical_crossentropy',metrics=['accuracy'])

# Step 8: Train the Model

In [10]:
# Train the ANN model

history = model.fit(
    X_train, y_train,
    epochs=30,             # Number of times model sees data
    batch_size=32,         # Data processed at once
    validation_split=0.2   # 20% of training used for validation
)

Epoch 1/30
40/40 ━━━━━━━━━━━━━━━━━━━━ 3s 16ms/step - accuracy: 0.2991 - loss: 1.3910 - val_accuracy: 0.3969 - val_loss: 1.3061
Epoch 2/30
40/40 ━━━━━━━━━━━━━━━━━━━━ 1s 7ms/step - accuracy: 0.4703 - loss: 1.2128 - val_accuracy: 0.5344 - val_loss: 1.1117
Epoch 3/30
40/40 ━━━━━━━━━━━━━━━━━━━━ 0s 8ms/step - accuracy: 0.6496 - loss: 0.9582 - val_accuracy: 0.6156 - val_loss: 0.8915
Epoch 4/30
40/40 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step - accuracy: 0.7279 - loss: 0.7628 - val_accuracy: 0.7219 - val_loss: 0.7038
Epoch 5/30
40/40 ━━━━━━━━━━━━━━━━━━━━ 1s 15ms/step - accuracy: 0.8182 - loss: 0.5946 - val_accuracy: 0.7875 - val_loss: 0.5808
Epoch 6/30
40/40 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step - accuracy: 0.8705 - loss: 0.4641 - val_accuracy: 0.8406 - val_loss: 0.4902
Epoch 7/30
40/40 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9006 - loss: 0.3749 - val_accuracy: 0.8719 - val_loss: 0.4229
Epoch 8/30
40/40 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - accuracy: 0.9161 - loss: 0.3155 - val_accuracy: 0.8938 - val_los

# Analysis on above result,
1. Here the Accuracy shows as 0.9995 which is really good
2. Loss shows a very low value of 0.0312

# Step 9: Evaluate the Model

In [11]:
# Evaluate model performance

loss, accuracy = model.evaluate(X_test, y_test)

print("Test Accuracy:", accuracy)

13/13 ━━━━━━━━━━━━━━━━━━━━ 0s 8ms/step - accuracy: 0.9428 - loss: 0.1599
Test Accuracy: 0.925000011920929


# Step 10: Make Predictions

In [12]:
# Make predictions

y_pred = model.predict(X_test)

# Convert probabilities to class labels
y_pred_classes = np.argmax(y_pred, axis=1)

print("Accuracy Score:", accuracy_score(y_test, y_pred_classes))
print(classification_report(y_test, y_pred_classes))

13/13 ━━━━━━━━━━━━━━━━━━━━ 0s 25ms/step
Accuracy Score: 0.925
              precision    recall  f1-score   support

           0       0.96      0.95      0.96       105
           1       0.89      0.91      0.90        91
           2       0.87      0.90      0.89        92
           3       0.96      0.93      0.95       112

    accuracy                           0.93       400
   macro avg       0.92      0.92      0.92       400
weighted avg       0.93      0.93      0.93       400



# Step 11: Hyperparameter Tuning

In [13]:
# Tuned model
# Change:
#    Number of neurons (64, 128)
#    Epochs (50)
#    Learning rate (0.0001)

model = Sequential()
model.add(Dense(128, activation='relu', input_dim=X_train.shape[1]))
model.add(Dense(64, activation='relu'))
model.add(Dense(4, activation='softmax'))

model.compile(
    optimizer=Adam(learning_rate=0.0001),
    loss='sparse_categorical_crossentropy',
    metrics=['accuracy']
)

model.fit(X_train, y_train, epochs=50, batch_size=32)

/usr/local/lib/python3.12/dist-packages/keras/src/layers/core/dense.py:93: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 1/50
50/50 ━━━━━━━━━━━━━━━━━━━━ 2s 6ms/step - accuracy: 0.2644 - loss: 1.3917
Epoch 2/50
50/50 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - accuracy: 0.3080 - loss: 1.3585
Epoch 3/50
50/50 ━━━━━━━━━━━━━━━━━━━━ 1s 12ms/step - accuracy: 0.4065 - loss: 1.2954
Epoch 4/50
50/50 ━━━━━━━━━━━━━━━━━━━━ 1s 5ms/step - accuracy: 0.4729 - loss: 1.2473
Epoch 5/50
50/50 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - accuracy: 0.5189 - loss: 1.2040
Epoch 6/50
50/50 ━━━━━━━━━━━━━━━━━━━━ 1s 5ms/step - accuracy: 0.5591 - loss: 1.1540
Epoch 7/50
50/50 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - accuracy: 0.5996 - loss: 1.0969
Epoch 8/50
50/50 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - accuracy: 0.6339 - loss: 1.0376
Epoch 9/50
50/50 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.6590 - loss: 0.9907
Epoch 10/50
50/50 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.6761 - loss: 0.9551
Epoch 11/50
50/50 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - accuracy: 0.6910 - loss: 0.9042
Epoch 12/50
50/50 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.7234 - l

In [14]:
# Evaluate model performance afetr Hyperparameter Tuning

loss, accuracy = model.evaluate(X_test, y_test)

print("Test Accuracy:", accuracy)

13/13 ━━━━━━━━━━━━━━━━━━━━ 0s 10ms/step - accuracy: 0.9387 - loss: 0.2358
Test Accuracy: 0.9175000190734863


# Final Analsysis after performing the Hyperparameter Tuning,
Hyperparameter tuning was performed by adjusting parameters such as the number of neurons, learning rate, and number of epochs. The tuned model achieved an accuracy of 0.917, while the initial model achieved 0.925. This indicates that the original model configuration was already performing well for this dataset. Therefore, the initial model was retained as the better performing model.